backend/sqlite/sqlite.go

```diff
@@ -441,40 +444,75 @@ func (be *sqliteBackend) createOrchestrationInstanceInternal(ctx context.Context
 		return errors.New("HistoryEvent must be an ExecutionStartedEvent")
 	}
 
-	// TODO: Support for re-using orchestration instance IDs
-	res, err := tx.ExecContext(
+	var instanceExists bool
+	instanceRes, existErr := tx.QueryContext(
 		ctx,
-		`INSERT OR IGNORE INTO [Instances] (
-			[Name],
-			[Version],
-			[InstanceID],
-			[ExecutionID],
-			[Input],
-			[RuntimeStatus],
-			[CreatedTime]
-		) VALUES (?, ?, ?, ?, ?, ?, ?)`,
-		startEvent.Name,
-		startEvent.Version.GetValue(),
+		`SELECT * FROM [Instances] WHERE [InstanceID] = ?`, 
 		startEvent.OrchestrationInstance.InstanceId,
-		startEvent.OrchestrationInstance.ExecutionId.GetValue(),
-		startEvent.Input.GetValue(),
-		"PENDING",
-		e.Timestamp.AsTime(),
 	)
-	if err != nil {
-		return fmt.Errorf("failed to insert into [Instances] table: %w", err)
+	if existErr != nil {
+		return fmt.Errorf("failed to query from [Instances] table: %w", existErr)
 	}
 
-	rows, err := res.RowsAffected()
-	if err != nil {
-		return fmt.Errorf("failed to count the rows affected: %w", err)
-	}
+	if instanceRes == nil {
+		instanceExists = false
+	} else {
+		instanceExists = true
+	}
+
+	if !instanceExists || startEvent.InstanceExistOption == protos.InstanceExistOption_Throw_IF_EXIST {
+		// TODO: Support for re-using orchestration instance IDs
+		res, err := tx.ExecContext(
+			ctx,
+			`INSERT OR IGNORE INTO [Instances] (
+				[Name],
+				[Version],
+				[InstanceID],
+				[ExecutionID],
+				[Input],
+				[RuntimeStatus],
+				[CreatedTime]
+			) VALUES (?, ?, ?, ?, ?, ?, ?)`,
+			startEvent.Name,
+			startEvent.Version.GetValue(),
+			startEvent.OrchestrationInstance.InstanceId,
+			startEvent.OrchestrationInstance.ExecutionId.GetValue(),
+			startEvent.Input.GetValue(),
+			"PENDING",
+			e.Timestamp.AsTime(),
+		)
+		if err != nil {
+			return fmt.Errorf("failed to insert into [Instances] table: %w", err)
+		}
 
-	if rows <= 0 {
-		return backend.ErrDuplicateEvent
-	}
+		rows, err := res.RowsAffected()
+		if err != nil {
+			return fmt.Errorf("failed to count the rows affected: %w", err)
+		}
+
+		if rows <= 0 {
+			return backend.ErrDuplicateEvent
+		}
 
-	*instanceID = startEvent.OrchestrationInstance.InstanceId
+		*instanceID = startEvent.OrchestrationInstance.InstanceId
+	} else if startEvent.InstanceExistOption == protos.InstanceExistOption_SKIP_IF_EXIST {
+		be.logger.Debugf(
+			"InstanceExistOption_SKIP_IF_EXIST set, skip creating Instance with instanceId (%v), already exists",
+			startEvent.OrchestrationInstance.InstanceId)
+			// do nothing
+			return backend.ErrSkipExistInstance
+	} else {
+		// same instance id exists and InstanceExistOption_TERMINATE_IF_EXIST set
+		// to terminate old instance
+		// 1. remove all entires in history table
+		// 2. remove all entries in instance table
+		// 3. remove all entries in newEvents table
+		// 4. remove all entries in new task table
+		// 5. insert new instance to instance table. 
+		// do all above updates in one atomic operation
```

Comment:
We're already operating in the context of a transaction, so it's okay if you do this in multiple discrete steps.